domain/runaway.go

```diff
@@ -0,0 +1,525 @@
+		return
+	}
+	tb, err := cache.NewBasePhysicalTable(systemSchemaCIStr, tbInfo, model.NewCIStr(""), col)
+	if err != nil {
+		logutil.BgLogger().Error("delete system table failed", zap.String("table", tableName), zap.Error(err))
+		return
+	}
+	generator, err := sqlbuilder.NewScanQueryGenerator(tb, expiredTime, nil, nil)
+	if err != nil {
+		logutil.BgLogger().Error("delete system table failed", zap.String("table", tableName), zap.Error(err))
+		return
+	}
+	var leftRows [][]types.Datum
+	for {
+		sql := ""
+		if sql, err = generator.NextSQL(leftRows, runawayRecordGCSelectBatchSize); err != nil {
+			logutil.BgLogger().Error("delete system table failed", zap.String("table", tableName), zap.Error(err))
+			return
+		}
+		// to remove
+		if len(sql) == 0 {
+			return
+		}
+
+		rows, sqlErr := do.execRestrictedSQL(sql, nil)
+		if sqlErr != nil {
+			logutil.BgLogger().Error("delete system table failed", zap.String("table", tableName), zap.Error(err))
+			return
+		}
+		leftRows = make([][]types.Datum, len(rows))
+		for i, row := range rows {
+			leftRows[i] = row.GetDatumRow(tb.KeyColumnTypes)
+		}
+
+		for len(leftRows) > 0 {
+			var delBatch [][]types.Datum
+			if len(leftRows) < runawayRecordGCBatchSize {
+				delBatch = leftRows
+				leftRows = nil
+			} else {
+				delBatch = leftRows[0:runawayRecordGCBatchSize]
+				leftRows = leftRows[runawayRecordGCBatchSize:]
+			}
+			sql, err := sqlbuilder.BuildDeleteSQL(tb, delBatch, expiredTime)
+			if err != nil {
+				logutil.BgLogger().Error(
+					"build delete SQL failed when deleting system table",
+					zap.Error(err),
+					zap.String("table", tb.Schema.O+"."+tb.Name.O),
+				)
+				return
+			}
+
+			_, err = do.execRestrictedSQL(sql, nil)
+			if err != nil {
+				logutil.BgLogger().Error(
+					"delete SQL failed when deleting system table", zap.Error(err), zap.String("SQL", sql),
+				)
+			}
+		}
+	}
+}
+
+func (do *Domain) runawayWatchSyncLoop() {
+	defer util.Recover(metrics.LabelDomain, "runawayWatchSyncLoop", nil, false)
+	runawayWatchSyncTicker := time.NewTicker(runawayWatchSyncInterval)
+	updateFn := func() error {
+		records, err := do.runawaySyncer.getNewWatchRecords()
+		if err != nil {
+			logutil.BgLogger().Error("try to get new runaway watch", zap.Error(err))
+			return err
+		}
+		for _, r := range records {
+			do.runawayManager.AddWatch(r)
+		}
+		doneRecords, err := do.runawaySyncer.getNewWatchDoneRecords()
+		if err != nil {
+			logutil.BgLogger().Error("try to get done runaway watch", zap.Error(err))
+			return err
+		}
+		for _, r := range doneRecords {
+			do.runawayManager.RemoveWatch(r)
+		}
+		return nil
+	}
+	for {
+		select {
+		case <-do.exit:
+			return
+		case <-do.runawaySyncer.notifyChan:
+			err := updateFn()
+			do.runawaySyncer.doneChan <- err
+		case <-runawayWatchSyncTicker.C:
+			updateFn()
+		}
+	}
+}
+
+func (do *Domain) AddRunawayWatch(record *resourcegroup.QuarantineRecord) (int64, error) {
+	if err := do.handleRunawayWatch(record); err != nil {
```

Comment:
I think addwatch should return success after the record is inseted into the watch table. No need to wait for the background task refresh.  BTW, if a statement returns error, you should ensure there is no side effect in the system.

Reply:
fix

Reply:
In the design, we want to return the ID of the added watch record to the user.

Reply:
Then how do you handle the case if the record did insert successfully but the wait watch timeout, in your current impl it will return an error but because the record is inserted it will still take effect later and on other instances. Please ensure if you return an error to the user(which means the txn is not successful) then there should be no unexpected side-effect in the system.

Reply:
It should be a very rare situation. I think there are no side effects, but we should return a clearer error message that the user added successfully, but the TiDB server load timed out, you can slightly query information_schame.runaway_watches to confirm.

Reply:
Don't get what you want to explain, but it's still clear that the current logic cannot ensure the procedure to be atomic, so it can deliver wrong result

Reply:
Well, I know what you mean. The main point of disagreement is whether the function needs to keep atomic. I didn't think we needed it, so I didn't implement it strictly.